ocaml/xapi/repository_helpers.ml

```diff
@@ -0,0 +1,732 @@
+    | "gte" -> [Gt; Eq]
+    | "lte" -> [Lt; Eq]
+    | "gt"  -> [Gt]
+    | "lt"  -> [Lt]
+    | "eq"  -> [Eq]
+    | _ -> raise Invalid_inequalities
+
+  let string_of_inequalities = function
+    | [Gt; Eq] | [Eq; Gt] -> ">="
+    | [Lt; Eq] | [Eq; Lt] -> "<="
+    | [Gt] -> ">"
+    | [Lt] -> "<"
+    | [Eq] -> "="
+    | _ -> raise Invalid_inequalities
+
+  let default = {
+      name = ""
+    ; arch = ""
+    ; inequalities = []
+    ; epoch = ""
+    ; version = ""
+    ; release = ""
+    }
+
+  let of_xml = function
+    | Xml.Element ("applicability", _, children) ->
+      List.fold_left (fun a n ->
+          match n with
+          | Xml.Element ("inequality", _, [Xml.PCData v]) ->
+            { a with inequalities = (inequalities_of_string v) }
+          | Xml.Element ("epoch", _, [Xml.PCData v]) ->
+            { a with epoch = v }
+          | Xml.Element ("version", _, [Xml.PCData v]) ->
+            { a with version = v }
+          | Xml.Element ("release", _, [Xml.PCData v]) ->
+            { a with release = v }
+          | Xml.Element ("name", _, [Xml.PCData v]) ->
+            { a with name = v }
+          | Xml.Element ("arch", _, [Xml.PCData v]) ->
+            { a with arch = v }
+          | _ ->
+            error "Unknown node in <applicability>";
+            raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+      ) default children
+    | _ ->
+      error "Unknown node in <guidance_applicabilities>";
+      raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+
+  let to_string a =
+    Printf.sprintf "%s%s%s-%s"
+      a.name (string_of_inequalities a.inequalities) a.version a.release
+
+  let compare_version_strings s1 s2 =
+    (* Compare versions or releases of RPM packages
+     * I.E. for "libpath-utils-0.2.1-29.el7.x86_64" and "libpath-utils-0.2.1a-30.el7.x86_64",
+     * this function compares:
+     * versions between "0.2.1" and "0.2.1a", or
+     * releases between "29.el7" and "30.el7".
+     * More examples:
+     *  "1.2.3" "<" "1.2.4"
+     *  "1.2.3" "=" "1.2.3"
+     *  "1.2.3" ">" "1.2"
+     *  "1.0011" ">" "1.9"
+     *  "1.05" "=" "1.5"
+     *  "1.0" ">" "1"
+     *  "1.0" ">" "1.a"
+     *  "2.50" ">" "2.5"
+     *  "XS3" "<" "xs2"
+     *  "1.2.3" ">" "1.2.3a"
+     *  "xs4" "=" "xs.4"
+     *  "2a" "<" "2.0"
+     *  "2a" "<" "2b"
+     *  "1.0" ">" "1.xs2"
+     *  "1.0_xs" "=" "1.0.xs"
+     *)
+    let normalize v =
+      let split_letters_and_numbers s =
+        let r = Re.Str.regexp "^\\([^0-9]+\\)\\([0-9]+\\)$" in
+        if Re.Str.string_match r s 0 then
+          [(Re.Str.replace_first r "\\1" s); (Re.Str.replace_first r "\\2" s)]
+        else [s]
+      in
+      let concat_map f l = List.concat (List.map f l) in
+      v
+      |> Astring.String.cuts ~sep:"."
+      |> concat_map (fun s -> Astring.String.cuts ~sep:"_" s)
+      |> concat_map (fun s -> split_letters_and_numbers s)
+      |> List.map (fun s -> try Int (int_of_string s) with _ -> Str s)
```

Comment:
Is this new code? The detection of integers is not safe:

```
utop # int_of_string "0x10";;
- : int = 16
```

Reply:
Merged into feature branch already. 
This commit re-structure the code with modules and helpers.